core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -425,36 +425,35 @@ class Partition(val topicPartition: TopicPartition,
   }
 
   /**
-   * Checks if the topic ID provided in the request is consistent with the topic ID in the log.
+   * Checks if the topic ID received is consistent with the topic ID in the log.
    * If a valid topic ID is provided, and the log exists but has no ID set, set the log ID to be the request ID.
    *
-   * @param requestTopicId the topic ID from the request
-   * @return true if the request topic id is consistent, false otherwise
+   * @param receivedTopicId the topic ID from the LeaderAndIsr request or from the metadata records
+   * @return true if the received topic id is consistent, false otherwise
    */
-  def checkOrSetTopicId(requestTopicId: Uuid): Boolean = {
-    // If the request had an invalid topic ID, then we assume that topic IDs are not supported.
-    // The topic ID was not inconsistent, so return true.
-    // If the log is empty, then we can not say that topic ID is inconsistent, so return true.
-    if (requestTopicId == null || requestTopicId == Uuid.ZERO_UUID)
-      true
-    else {
+  def checkOrSetTopicId(receivedTopicId: Uuid, usingRaft: Boolean): Boolean = {
```

Comment:
I think the path through here makes sense when we're talking about the old zk logic in which we may or may not have an existing topicId and we rely on the controller to tell us. However, for the quorum controller, we should always have the topic ID when we create the `Log` object, so it feels strange to need to update it. That makes me think we should be looking at the paths that lead to `LogManager.getOrCreateLog`. Does it make sense to pass through topicId at that point?

Reply:
Yeah. We don't even "update" in the ZK code path, but we do wait to create the file. If we have the topic ID available when we first create the log, it would make more sense to write the file at that point. The code path in RaftReplicaManager could just be checking the log.

Reply:
Hmm. Looking into this further, it seems that the log is created in RaftReplicaManager when making a leader or a follower. This occurs in the `handleMetadataRecords` method and the `endMetadataChangeDeferral` method.  Each requires a few methods to pass through before we get to `LogManager.getOrCreateLog`, which is not ideal to pass a topic ID through (especially when some of this code is shared with the ZK path, but not impossible)

One thing I am wondering though is how I tested the changes. I think I assume the log is created already, but this code suggests that the log will not be created until `makeLeaders/Followers` in the non deferring case, and not created until `endMetadataChangeDeferral` in the deferring case. This means that where I set topic IDs now, we won't have a log on the first pass! So I think maybe it is worth it to pass the ID through these methods. I think I can keep the check of the topic ID where it is though.

Reply:
Hmm this is a little tricky. One idea I had would be to check or set the ID in the partition object only (not the file) then when the the partitions are passed in to `makeLeaders/Followers` and a new log is created, we will have the topic ID easily accessible to create the partition.metadata file. It kind of changes how we do things. We would either store in partition _instead of_ or _in addition to_  Log. Not sure if we also want to change this for the ZK code. Thoughts?

Reply:
Hmmm I'm not a huge fan of this either since it also seems to be waiting to set the ID and it adds complexity to recovery. I'll take another look. Maybe I'm missing something.